logstash-core/lib/logstash/output_delegator.rb

```diff
@@ -1,192 +1,60 @@
+require "logstash/output_delegator_strategy_registry"
+
+require "logstash/output_delegator_strategies/shared"
+require "logstash/output_delegator_strategies/single"
 
-# This class goes hand in hand with the pipeline to provide a pool of
-# free workers to be used by pipeline worker threads. The pool is
-# internally represented with a SizedQueue set the the size of the number
-# of 'workers' the output plugin is configured with.
-#
-# This plugin also records some basic statistics
 module LogStash class OutputDelegator
-  attr_reader :workers, :config, :threadsafe
+  attr_reader :metric, :metric_events, :strategy, :namespaced_metric, :metric_events , :plugin_args
 
-  # The *args this takes are the same format that a Outputs::Base takes. A list of hashes with parameters in them
-  # Internally these just get merged together into a single hash
-  def initialize(logger, klass, default_worker_count, metric, *plugin_args)
+  def initialize(logger, output_class, metric, xopts={}, plugin_args={})
     @logger = logger
-    @threadsafe = klass.threadsafe?
-    @config = plugin_args.reduce({}, :merge)
-    @klass = klass
-    @workers = java.util.concurrent.CopyOnWriteArrayList.new
-    @default_worker_count = default_worker_count
-    @registered = false
-
-    # Create an instance of the input so we can fetch the identifier
-    output = @klass.new(@config)
-
-    # Scope the metrics to the plugin
-    namespaced_metric = metric.namespace(output.plugin_unique_name.to_sym)
-    output.metric = namespaced_metric
+    @output_class = output_class
+    @metric = metric
+    @xopts = xopts # Options specific to the given class
+    @plugin_args = plugin_args
+    build_strategy!
 
-    @metric_events = namespaced_metric.namespace(:events)
-    namespaced_metric.gauge(:name, config_name)
-
-    @events_received = Concurrent::AtomicFixnum.new(0)
+    @id = plugin_args["id"]
+    @namespaced_metric = metric.namespace("#{@output_class.config_name}_#{@id}".to_sym)
+    @metric_events = @namespaced_metric.namespace(:events)
+    @namespaced_metric.gauge(:name, config_name)
   end
 
-  def threadsafe?
-    !!@threadsafe
+  def concurrency
+    @output_class.concurrency
   end
 
-  def warn_on_worker_override!
-    # The user has configured extra workers, but this plugin doesn't support it :(
-    if worker_limits_overriden?
-      message = @klass.workers_not_supported_message
-      warning_meta = {:plugin => @klass.config_name, :worker_count => @config["workers"]}
-      if message
-        warning_meta[:message] = message
-        @logger.warn(I18n.t("logstash.pipeline.output-worker-unsupported-with-message", warning_meta))
-      else
-        @logger.warn(I18n.t("logstash.pipeline.output-worker-unsupported", warning_meta))
-      end
-    end
-  end
+  def build_strategy!
+    # TODO: Make this injected
+    @strategy_registry = LogStash::OutputDelegatorStrategyRegistry.instance
+    @strategy_class = @strategy_registry.get(self.concurrency)
```

Comment:
Consider raising the "Could not find output delegator strategy of type..." error in the `instance.get()` method (it is probably where it belongs, as any other method using it will have to also test and raise).
However, maybe `class_for` makes more sense .e.g. `.instance.class_for(self.concurrency).new(...)`
Gives...
```ruby
  def build_strategy!
    @output_delegate = LogStash::OutputDelegatorStrategyRegistry.instance
        .class_for(self.concurrency)
        .new(@logger, @output_class, @metric, @xopts, @plugin_args)
  end
```